src/transaction/atomic_replication_helper.cpp

```diff
@@ -165,25 +188,62 @@ namespace cublog
     PGBUF_CLEAR_WATCHER (&m_watcher);
   }
 
+  void
+  atomic_replication_helper::atomic_replication_sequence::atomic_replication_unit::apply_log_redo (THREAD_ENTRY *thread_p,
+      log_rv_redo_context &redo_context)
+  {
+    redo_context.m_reader.set_lsa_and_fetch_page (m_record_lsa, log_reader::fetch_mode::FORCE);
+    const log_rec_header header = redo_context.m_reader.reinterpret_copy_and_add_align<log_rec_header> ();
+
+    switch (header.type)
+      {
+      case LOG_REDO_DATA:
+	apply_log_by_type<LOG_REC_REDO> (thread_p, redo_context, header.type);
+	break;
+      case LOG_MVCC_REDO_DATA:
+	apply_log_by_type<LOG_REC_MVCC_REDO> (thread_p, redo_context, header.type);
+	break;
+      case LOG_UNDOREDO_DATA:
+      case LOG_DIFF_UNDOREDO_DATA:
+	apply_log_by_type<LOG_REC_UNDOREDO> (thread_p, redo_context, header.type);
+	break;
+      case LOG_MVCC_UNDOREDO_DATA:
+      case LOG_MVCC_DIFF_UNDOREDO_DATA:
+	apply_log_by_type<LOG_REC_MVCC_UNDOREDO> (thread_p, redo_context, header.type);
+	break;
+      case LOG_RUN_POSTPONE:
+	apply_log_by_type<LOG_REC_RUN_POSTPONE> (thread_p, redo_context, header.type);
+	break;
+      case LOG_COMPENSATE:
+	apply_log_by_type<LOG_REC_COMPENSATE> (thread_p, redo_context, header.type);
+	break;
+      default:
+	break;
```

Comment:
Assert that default should not happen since all log record types that need to be avoided are already avoided in the replicator.